jekyll/_posts/2021-03-12-new-Nim-channels.md

```diff
@@ -0,0 +1,224 @@
+
+### It is safe and convenient
+
+The Nim compiler rejects the program below at compile time. It says that `expression cannot be isolated: s`. This is because `s` is a `ref object` - it may be modified somewhere and is not unique, so the variable cannot be isolated.
+
+
+```nim
+import std/[channels, json, isolation]
+
+var chan = newChannel[JsonNode]()
+
+proc spawnCrawlers =
+  var s = newJString("Hello, Nim")
+  chan.send isolate(s)
+```
+
+It is only allowed to pass a function call directly.
+
+```nim
+import std/[channels, json, isolation]
+
+var chan = newChannel[JsonNode]()
+
+proc spawnCrawlers =
+  chan.send isolate(newJString("Hello, Nim"))
+```
+
+`Isolated` data can only be moved, not copied. It is implemented as a library without bloating Nim's core type system. The `isolate` proc is used to create an isolated subgraph from the expression `value`. Whether the expression `value` is isolated is checked at compile time. The `extract` proc is used to get the internal value of `Isolated` data.
+
+```nim
+import std/isolation
+
+var data = isolate("string")
+doAssert data.extract == "string"
+doAssert data.extract == ""
+```
+
+By means of `Isolated` data, the channels become safer and more convenient to use.
+
+
+## Benchmark
+
+Here is a simple benchmark. We create 10 threads that send data to the channel and 5 threads that receive it.
+
+```nim
+# benchmark the new channel implementation with 
+# `nim c -r --threads:on --gc:orc -d:newChan -d:danger app.nim`
+#
+# benchmark the old channel implementation with
+# `nim c -r --threads:on -d:oldChan -d:danger app.nim`
+
+import std/times
+
+var
+  sender: array[0 .. 9, Thread[void]]
+  receiver: array[0 .. 4, Thread[void]] 
+when defined(newChan):
+  import std/[channels, isolation]
+  var chan = newChannel[seq[string]](40)
+
+  proc sendHandler() =
+    chan.send(isolate(@["Hello, Nim"]))
+
+  proc recvHandler() =
+    var x: seq[string]
+    chan.recv(x)
+    discard x
+
+elif defined(oldChan):
+  var chan: Channel[seq[string]]
+
+  chan.open(maxItems = 40)
+
+  proc sendHandler() =
+    chan.send(@["Hello, Nim"])
+  proc recvHandler() =
+    let x = chan.recv()
+    discard x
+
+template benchmark() =
+  for i in 0 .. sender.high:
+    createThread(sender[i], sendHandler)
+
+  joinThreads(sender)
+  for i in 0 .. receiver.high:
+    createThread(receiver[i], recvHandler)
+
+  let start = now()
+  joinThreads(receiver)
+  echo now() - start
+
+benchmark()
+```
+
+The new implementation is much faster than the old one!
+
+
+| Implementation                     | Elapsed time |
+| ---------------------------------- | -----------: |
+| system/channels + refc (-d:danger) |       433 μs |
+| std/channels + orc (-d:danger)     |       137 μs |
```

Comment:
* use `system/channels_builtin` (that PR was already merged)
* not apples to apples comparison, you'd also need to compare with:
system/channels_builtin + orc (-d:danger)

Reply:
a bit improvement